src/client/QXmppOmemoManager.cpp

```diff
@@ -0,0 +1,5446 @@
+
+        // last part of hkdfKey
+        auto initializationVector = QCA::InitializationVector(PAYLOAD_INITIALIZATION_VECTOR_SIZE);
+        const auto initializationVectorOffset = hkdfOutput.data() + PAYLOAD_KEY_SIZE + PAYLOAD_AUTHENTICATION_KEY_SIZE;
+        std::copy(initializationVectorOffset, initializationVectorOffset + PAYLOAD_INITIALIZATION_VECTOR_SIZE, initializationVector.data());
+
+        QCA::Cipher cipher(PAYLOAD_CIPHER_TYPE, PAYLOAD_CIPHER_MODE, PAYLOAD_CIPHER_PADDING, QCA::Encode, encryptionKey, initializationVector);
+        auto encryptedPayload = cipher.process(QCA::MemoryRegion(payload));
+
+        if (encryptedPayload.isEmpty()) {
+            q->warning("Following payload could not be encrypted: " % QString::fromUtf8(payload));
+            return {};
+        }
+
+        if (!QCA::MessageAuthenticationCode::supportedTypes().contains(PAYLOAD_MESSAGE_AUTHENTICATION_CODE_TYPE)) {
+            q->warning("Message authentication code type '" % QString(PAYLOAD_MESSAGE_AUTHENTICATION_CODE_TYPE) % "' is not supported by this system");
+            return {};
+        }
+
+        auto messageAuthenticationCodeGenerator = QCA::MessageAuthenticationCode(PAYLOAD_MESSAGE_AUTHENTICATION_CODE_TYPE, authenticationKey);
+        auto messageAuthenticationCode = QCA::SecureArray(messageAuthenticationCodeGenerator.process(encryptedPayload));
+        messageAuthenticationCode.resize(PAYLOAD_MESSAGE_AUTHENTICATION_CODE_SIZE);
+
+        PayloadEncryptionResult payloadEncryptionData;
+        payloadEncryptionData.decryptionData = hkdfKey.append(messageAuthenticationCode);
+        payloadEncryptionData.encryptedPayload = encryptedPayload.toByteArray();
+
+        return payloadEncryptionData;
+    }
+
+    ///
+    /// Creates the SCE envelope as defined in
+    /// \xep{0420, Stanza Content Encryption} for a message or IQ stanza.
+    ///
+    /// The stanza's content that should be enrcypted is put into the SCE
+    /// content and that is added to the SCE envelope.
+    /// Additionally, the standard SCE affix elements are added to the SCE
+    /// envelope.
+    ///
+    /// \param stanza stanza for whom the SCE envelope is created
+    ///
+    /// \return the serialized SCE envelope
+    ///
+    template<typename T>
+    QByteArray createSceEnvelope(const T &stanza)
+    {
+        const auto randomGenerator = QRandomGenerator::system();
+        const auto sceRpadSize = randomGenerator->bounded(SCE_RPAD_SIZE_MIN, SCE_RPAD_SIZE_MAX);
+        QByteArray sceRpad;
+
+        // sceRpad is filled with unsigned integers.
+        // Each of them represents four characters.
+        // Thus, only (sceRpadSize / 4) unsigned integers are randomly generated.
+        sceRpad.resize(sceRpadSize);
+        randomGenerator->fillRange(reinterpret_cast<unsigned int *>(sceRpad.data()), sceRpadSize / sizeof(unsigned int));
+        sceRpad = sceRpad.toBase64();
```

Comment:
randomly the last 0-3 bytes are not initialized, because sceRpadSize is in bytes and only `sceRpadSize / sizeof(unsigned int)` full ints are written to the byte array.